enterprise/internal/licensing/check.go

```diff
@@ -0,0 +1,138 @@
+package licensing
+
+import (
+	"bytes"
+	"encoding/hex"
+	"encoding/json"
+	"io"
+	"time"
+
+	"net/http"
+
+	"github.com/sourcegraph/log"
+
+	"context"
+
+	"github.com/sourcegraph/sourcegraph/internal/conf"
+	"github.com/sourcegraph/sourcegraph/internal/goroutine"
+	"github.com/sourcegraph/sourcegraph/internal/httpcli"
+	"github.com/sourcegraph/sourcegraph/internal/redispool"
+	"github.com/sourcegraph/sourcegraph/lib/errors"
+)
+
+var (
+	licenseCheckStarted = false
+	store               = redispool.Store
+
+	lastCalledAtStoreKey    = "licensing:last_called_at"
+	licenseValidityStoreKey = "licensing:is_license_valid"
+	prevLicenseTokenKey     = "licensing:prev_license_hash"
+
+	routine *goroutine.PeriodicGoroutine
+)
+
+type licenseChecker struct {
+	siteID string
+	token  string
+	info   *Info
+	doer   httpcli.Doer
+}
+
+func (l *licenseChecker) Handle(ctx context.Context) error {
+	store.Set(lastCalledAtStoreKey, time.Now().Format(time.RFC3339))
+
+	if l.info.HasTag(AllowAirGappedTag) || l.info.SalesforceSubscriptionID != nil {
+		store.Set(licenseValidityStoreKey, true)
+		return nil
+	}
+
+	payload, err := json.Marshal(struct {
+		ClientSiteID string `json:"siteID"`
+	}{ClientSiteID: l.siteID})
+
+	if err != nil {
+		return err
+	}
+
+	req, err := http.NewRequest(http.MethodPost, "https://sourcegraph.com/.api/license/check", bytes.NewBuffer(payload))
+	if err != nil {
+		return err
+	}
+
+	req.Header.Set("Authorization", "Bearer "+l.token)
+	req.Header.Set("Content-Type", "application/json")
+
+	res, err := l.doer.Do(req)
+	if err != nil {
+		return err
+	}
+	defer res.Body.Close()
+	if res.StatusCode != http.StatusOK {
+		return errors.Newf("Failed to check license, status code: %d", res.StatusCode)
+	}
+
+	var body LicenseCheckResponse
+	resBody, err := io.ReadAll(res.Body)
+	if err != nil {
+		return err
+	}
+
+	json.Unmarshal([]byte(resBody), &body)
```

Comment:
You can make this a bit nicer by doing:

```suggestion
	var body LicenseCheckResponse
	if err := json.NewDecoder(res.Body).Decode(&body); err != nil {
		return err
	}
```

Then it does the Unmarshal and body reading error checking all in one

Reply:
TIL: This is definitely way more cleaner, thank you! 👍